### ISSUE TITLE
ambiguous regression

### ISSUE THREAD

**Issue by alephzero** [nlohmann/json#3204]
There was a regression between `v3.10.3` and `v3.10.4`, in the compiler's ability to disambiguate `std::function` containing `nlohmann::json`.

In compiler explorer:
```cpp
#include <https://raw.githubusercontent.com/nlohmann/json/v3.10.3/single_include/nlohmann/json.hpp>
#include <functional>
#include <string>

struct Foo {
  Foo() {}
  Foo(std::string) {}
};

struct Bar {
  Bar(std::function<void(Foo)>) {}
  Bar(std::function<void(nlohmann::json)>) {}
};

int main() {
  Bar bar([](Foo) {});
}
```
Generates
```
ASM generation compiler returned: 0
Execution build compiler returned: 0
Program returned: 0
```

Changing the version to `v3.10.4`:
```cpp
#include <https://raw.githubusercontent.com/nlohmann/json/v3.10.4/single_include/nlohmann/json.hpp>
#include <functional>
#include <string>

struct Foo {
  Foo() {}
  Foo(std::string) {}
};

struct Bar {
  Bar(std::function<void(Foo)>) {}
  Bar(std::function<void(nlohmann::json)>) {}
};

int main() {
  Bar bar([](Foo) {});
}
```
Generates
```
<source>: In function 'int main()':
<source>:16:21: error: call of overloaded 'Bar(main()::<lambda(Foo)>)' is ambiguous
<source>:12:3: note: candidate: 'Bar::Bar(std::function<void(nlohmann::basic_json<>)>)'
<source>:11:3: note: candidate: 'Bar::Bar(std::function<void(Foo)>)'
ASM generation compiler returned: 1
<source>: In function 'int main()':
<source>:16:21: error: call of overloaded 'Bar(main()::<lambda(Foo)>)' is ambiguous
<source>:12:3: note: candidate: 'Bar::Bar(std::function<void(nlohmann::basic_json<>)>)'
<source>:11:3: note: candidate: 'Bar::Bar(std::function<void(Foo)>)'
Execution build compiler returned: 1
```

**Comment by alephzero** [comment]
This is still an issue in `v3.10.5`

**Comment by falbrechtskirchinger** [comment]
Like #3384 this one is also caused by 0e694b4060ed55df980eaaebc2398b0ff24530d4.

**Comment by falbrechtskirchinger** [comment]
@alephzero Unfortunately I don't believe your issue is fixable in the short term. Above mentioned commit enabled the intended behavior of implicit string conversion so your only options are to either turn of implicit conversions (define `JSON_USE_IMPLICIT_CONVERSIONS=0`) or de-rank an overload in the ambiguous overload set, e.g.:

```cpp
#include <https://raw.githubusercontent.com/nlohmann/json/develop/single_include/nlohmann/json.hpp>
#include <functional>
#include <string>

struct Foo {
  Foo() {}
  Foo(std::string) {}
};

struct Bar {
  // de-rank overload by making it a template
  template<typename>
  Bar(std::function<void(Foo)>) {}
  Bar(std::function<void(nlohmann::json)>) {}
};

int main() {
  Bar bar([](Foo) {});
}
```

@nlohmann Thoughts?

**Comment by falbrechtskirchinger** [comment]
@nlohmann This issue was fixed by #3604.
